### PR TITLE
feat(language-service): add code reactoring action to migrate `@Input` to signal input

### PR DESCRIPTION
**Pull request by devversion**
(experimental action for now; migration is not done/finalized yet)

Language service refactoring action that can convert `@Input()` declarations to signal inputs.

The user can click on an `@Input` property declaration in e.g. the VSCode extension and ask
for the input to be migrated. All references, imports and the declaration are updated automatically.

**NOTE**: Currently based on top of: #57562

### CODE REVIEW

**Review by crisbeto — APPROVED** [review]

File: packages/language-service/src/language_service.ts [hunk 1]
```diff
@@ -549,12 +549,12 @@ export class LanguageService {
     );
   }
 
-  applyRefactoring(
+  async applyRefactoring(
```
Comment (atscott) [inline packages/language-service/src/language_service.ts:552]: Would you mind adding a comment here about the ramifications of making this async (basically what's in the commit message)

File: packages/core/schematics/migrations/signal-migration/src/migration.ts [hunk 2]
```diff
@@ -38,6 +38,12 @@ export class SignalInputMigration extends TsurgeComplexMigration<
   upgradeAnalysisPhaseToAvoidBatch = false;
   upgradedAnalysisPhaseResults: Replacement[] | null = null;
 
+  // Necessary for language service configuration.
+  lsReportProgressFn: ((percentage: number, updateMessage: string) => void) | null = null;
```
Comment (atscott) [inline packages/core/schematics/migrations/signal-migration/src/migration.ts:42]:
```suggestion
  reportProgressFn: ((percentage: number, updateMessage: string) => void) | null = null;
```

File: packages/language-service/src/refactorings/convert_to_signal_input.ts [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+  while (!ts.isPropertyDeclaration(node) && !ts.isSourceFile(node)) {
+    node = node.parent;
+  }
+  if (ts.isSourceFile(node)) {
+    return null;
+  }
+  return node;
+}
+
+function getBeforeMigrateHookToFilterAllUnrelatedInputs(
```
Comment (atscott) [inline packages/language-service/src/refactorings/convert_to_signal_input.ts:200]: This is really difficult to understand because of how it mutates the known inputs and results prior to running the migration on them. It's really hard to follow that this is happening when looking at the actual migration code that calls the hook: from there, it looks like the hook might just be a no-op for the migration since it's a void function. Can we change this to have the "beforeMigration" actually return an object with the modified known inputs and results?

Alternatively, can we document this "beforeMigrateHook" either by renaming it to something that indicates the objects being passed in can be/are mutated or add some jsdoc to the function or comments at the call location in migration.ts?

File: packages/language-service/src/refactorings/convert_to_signal_input.ts [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    // Note that we still analyzed the whole application for potential references.
+    // Only migrate references to the target input.
+    Array.from(result.sourceInputs.keys()).forEach(
+      (i) =>
+        i.key !== key &&
+        knownInputs.markInputAsIncompatible(i, {
+          context: null,
+          reason: InputIncompatibilityReason.IgnoredBecauseOfLanguageServiceRefactoringRange,
+        }),
+    );
```
Comment (atscott) [inline packages/language-service/src/refactorings/convert_to_signal_input.ts:224]: nit: for..of https://google.github.io/styleguide/tsguide.html#iterating-containers

File: packages/language-service/src/refactorings/convert_to_signal_input.ts [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+    }
+
+    setTargetInput(targetInput);
+
+    // Mark all other inputs as incompatible.
+    // Note that we still analyzed the whole application for potential references.
+    // Only migrate references to the target input.
+    Array.from(result.sourceInputs.keys()).forEach(
+      (i) =>
+        i.key !== key &&
```
Comment (atscott) [inline packages/language-service/src/refactorings/convert_to_signal_input.ts:219]: nit: I feel like this would be easier to read as`if (i.key !== key) { ... }`

**Review by atscott — APPROVED** [review]

File: packages/language-service/src/language_service.ts (hunk 1 above)
Comment (devversion) [reply packages/language-service/src/language_service.ts:552]: good point. Done!

File: packages/language-service/src/refactorings/convert_to_signal_input.ts (hunk 4 above)
Comment (devversion) [reply packages/language-service/src/refactorings/convert_to_signal_input.ts:224]: ops. yeah.

File: packages/language-service/src/refactorings/convert_to_signal_input.ts (hunk 5 above)
Comment (devversion) [reply packages/language-service/src/refactorings/convert_to_signal_input.ts:219]: yeah, this is now much cleaner after getting rid of the `forEach` (which was a leftover from quick testing)

File: packages/language-service/src/refactorings/convert_to_signal_input.ts (hunk 3 above)
Comment (devversion) [reply packages/language-service/src/refactorings/convert_to_signal_input.ts:200]: It was done that way to keep the migration code more decoupled from the language service, and it's indeed leveraging a few implementation details.

You are right though. It doesn't feel great. I will have a look, but don't want to make super invasive changes right now since I also need to figure out the same problem when trying to limit to an directory in `ng update`.

Comment (devversion) [reply packages/language-service/src/refactorings/convert_to_signal_input.ts]: I will keep that comment open, and address it in a follow-up, tomorrow. Want to get things moving for now.

**Comment by AndrewKushnir** [comment]
This PR was merged into the repository by commit 82f45902163f89d64ee48109f54a1e75015d3af1.

The changes were merged into the following branches: main
